src/FrequencyTimeBalancer.cpp

```diff
@@ -0,0 +1,307 @@
+                int subdomain_group_count,
+                std::function<bool(int)> do_ignore_domain_idx,
+                double minimum_frequency,
+                double maximum_frequency,
+                std::unique_ptr<FrequencyLimitDetector> frequency_limit_detector);
+            virtual ~FrequencyTimeBalancerImp() = default;
+
+            std::vector<double> balance_frequencies_by_time(
+                const std::vector<double> &previous_times,
+                const std::vector<double> &previous_control_frequencies,
+                const std::vector<double> &previous_achieved_frequencies,
+                const std::vector<double> &previous_max_frequencies) override;
+
+            std::vector<double> get_target_times() const override;
+
+            double get_cutoff_frequency(unsigned int core) const override;
+
+            static std::unique_ptr<FrequencyTimeBalancer> make_unique(
+                double uncertainty_window_seconds,
+                int subdomain_group_count,
+                std::function<bool(int)> do_ignore_domain_idx,
+                double minimum_frequency,
+                double maximum_frequency,
+                std::unique_ptr<FrequencyLimitDetector> frequency_limit_detector);
+            static std::shared_ptr<FrequencyTimeBalancer> make_shared(
+                double uncertainty_window_seconds,
+                int subdomain_group_count,
+                std::function<bool(int)> do_ignore_domain_idx,
+                double minimum_frequency,
+                double maximum_frequency,
+                std::unique_ptr<FrequencyLimitDetector> frequency_limit_detector);
+
+        private:
+            double m_uncertainty_window_seconds;
+            int m_subdomain_group_count;
+            std::function<bool(int)> m_do_ignore_domain_idx;
+            double m_minimum_frequency;
+            double m_maximum_frequency;
+            std::vector<double> m_target_times;
+            std::vector<double> m_cutoff_frequencies;
+            std::unique_ptr<FrequencyLimitDetector> m_frequency_limit_detector;
+    };
+
+    FrequencyTimeBalancerImp::FrequencyTimeBalancerImp(
+        double uncertainty_window_seconds,
+        int subdomain_group_count,
+        std::function<bool(int)> do_ignore_domain_idx,
+        double minimum_frequency,
+        double maximum_frequency,
+        std::unique_ptr<FrequencyLimitDetector> frequency_limit_detector)
+        : m_uncertainty_window_seconds(uncertainty_window_seconds)
+        , m_subdomain_group_count(subdomain_group_count)
+        , m_do_ignore_domain_idx(do_ignore_domain_idx)
+        , m_minimum_frequency(minimum_frequency)
+        , m_maximum_frequency(maximum_frequency)
+        , m_target_times(subdomain_group_count, NAN)
+        , m_cutoff_frequencies(subdomain_group_count, NAN)
+        , m_frequency_limit_detector(std::move(frequency_limit_detector))
+    {
+    }
+
+    std::vector<double> FrequencyTimeBalancerImp::balance_frequencies_by_time(
```

Comment:
Suggestion: I feel this function is a bit lengthy and there's quite a bit going on, so I think it could be broken down into smaller functions to improve readability, allowing the reader to focus on more digestible pieces at a time.

Note: this is mostly coming from personal preference, as I'm used to smaller functions. To be fair, I've noticed our level of comfort regarding length of function is higher going by other functions I've seen in the codebase. So feel free to disregard the suggestion.